### ISSUE TITLE
[PPC] Support 64 bit code with 32 bit address space.

### ISSUE THREAD

**Issue by iMoD1998** [llvm/llvm-project#52826]
It would be great if someone would be kind enough to implement support for ppc64 based processors that use a 32 bit address space.

PS3 is an example of something that uses this and is currently a target (Lv2) but does not function correctly in both powerpc64 and powerpc targets.

It mostly breaks with function pointers with an example being:

```cpp
unsigned long long Example(unsigned long long x, unsigned long long y)
{
	return x * y;
}

auto ExamplePtr = (decltype(&Example))Example;

int main(int argc, const char* argv[])
{
	return ExamplePtr(1, 2);
}
```

Producing something along the lines of:
`Assertion failed: N1.getValueType() == N2.getValueType() && N1.getValueType() == VT && "Binary operator types must match!"`

Seems like this feature was intended to be implemented at one point as can have `use64BitRegs` feature?

Anyways, I would appreciate if anyone could add this or point me in the right direction to get started.

**Comment by luizfernandonb** [comment]
Any news on this?

**Comment by nemanjai** [comment]
I really doubt there is strong enough interest from anyone to invest the potentially significant amount of time required to implement this. I'm sorry to be the bearer of bad news.
Perhaps @lei137 can comment on the feasibility, but I'd be quite surprised to hear that someone was motivated enough.

**Comment by lei137** [comment]
You are right @nemanjai, I am not aware of any future plans to support this on my end as well.

**Comment by iMoD1998** [comment]
Its been awhile but from what i can remember the backend seems to already support this. PPC64 code in 32 bit is already supported by modifying/adding the cpu features.

https://github.com/llvm/llvm-project/blob/61a969b867573b892deef4a096fab87e89b13d98/llvm/lib/Target/PowerPC/PPC.td#L61

Feature64Bit/Feature64BitRegs
